# Проект по SQL

- **Опишите цели исследования**

Цели исследования:
Целью данного исследования является анализ базы данных, содержащей информацию о книгах, авторах, издательствах, пользовательских обзорах и оценках книг. В рамках этого анализа мы планируем выявить следующие аспекты:

1) Количество новых книг после 1 января 2000 года: Мы хотим определить, сколько книг было выпущено после указанной даты. Это позволит нам понять, насколько активно развивается рынок книг после появления интернет-технологий.

2) Обзоры и оценки книг: Нас интересует количество обзоров и средняя оценка для каждой книги. Эти данные помогут нам понять, какие книги пользуются наибольшей популярностью и востребованностью среди пользователей.

3) Издательства и объем выпущенных книг: Нашей целью также является определение издательства, которое выпустило наибольшее количество книг среди тех, которые имеют более 50 страниц. Это позволит выделить наиболее активных и влиятельных игроков на рынке.

4) Авторы с высокой оценкой: Мы хотим определить автора с самой высокой средней оценкой книг, учитывая только книги с достаточным количеством оценок. Это позволит нам выявить авторов, чьи произведения наиболее положительно оцениваются читателями.

5) Среднее количество обзоров для активных пользователей: Мы планируем подсчитать среднее количество обзоров от пользователей, которые оставили более 48 оценок. Это позволит нам понять, насколько активно пользователи взаимодействуют с контентом и вносят свой вклад в обсуждение книг.

Выполнение этих задач позволит нам получить ценную информацию о книжном рынке, предпочтениях пользователей и влиянии различных факторов на популярность книг.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

- **Исследуйте таблицы — выведите первые строки**

In [2]:
# Выведите первые строки из таблицы books
query_books = '''SELECT * FROM books LIMIT 5'''
books_data = pd.io.sql.read_sql(sql=text(query_books), con=engine)
print("Таблица books:")
display(books_data)

# Выведите первые строки из таблицы authors
query_authors = '''SELECT * FROM authors LIMIT 5'''
authors_data = pd.io.sql.read_sql(sql=text(query_authors), con=engine)
print("Таблица authors:")
display(authors_data)

# Выведите первые строки из таблицы publishers
query_publishers = '''SELECT * FROM publishers LIMIT 5'''
publishers_data = pd.io.sql.read_sql(sql=text(query_publishers), con=engine)
print("Таблица publishers:")
display(publishers_data)

# Выведите первые строки из таблицы ratings
query_ratings = '''SELECT * FROM ratings LIMIT 5'''
ratings_data = pd.io.sql.read_sql(sql=text(query_ratings), con=engine)
print("Таблица ratings:")
display(ratings_data)

# Выведите первые строки из таблицы reviews
query_reviews = '''SELECT * FROM reviews LIMIT 5'''
reviews_data = pd.io.sql.read_sql(sql=text(query_reviews), con=engine)
print("Таблица reviews:")
display(reviews_data)

Таблица books:


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Таблица authors:


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Таблица publishers:


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Таблица ratings:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Таблица reviews:


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

- **Сделайте по одному SQL-запросу для решения каждого задания**

**Задание 1: Посчитайте, сколько книг вышло после 1 января 2000 года.**

In [3]:
query = '''SELECT COUNT(*) AS num_books_after_2000 FROM books WHERE publication_date > '2000-01-01' '''
con = engine.connect()
result = pd.io.sql.read_sql(sql=text(query), con=con)

# Выводим результат
print("Задание 1:")
display(result)
print()

Задание 1:


num_books_after_2000
0                   819

*После 1 января 2000 года было выпущено 819 книг. Эта информация позволяет нам оценить активность издательств и авторов в новом тысячелетии.*

**Задание 2: Для каждой книги посчитайте количество обзоров и среднюю оценку.**

In [4]:
query = '''
SELECT
    b.title,
    COUNT(DISTINCT rv.review_id) AS num_reviews,
    AVG(ra.rating) AS avg_rating
FROM books b
LEFT JOIN reviews rv ON b.book_id = rv.book_id
LEFT JOIN ratings ra ON b.book_id = ra.book_id
GROUP BY b.book_id
ORDER BY b.book_id;
'''
result_2 = pd.io.sql.read_sql(sql=text(query), con=con)
print("Задание 2:")
display(result_2)
print()

Задание 2:


title  num_reviews  \
0                                         'Salem's Lot            2   
1                   1 000 Places to See Before You Die            1   
2    13 Little Blue Envelopes (Little Blue Envelope...            3   
3    1491: New Revelations of the Americas Before C...            2   
4                                                 1776            4   
..                                                 ...          ...   
995           Wyrd Sisters (Discworld  #6; Witches #2)            3   
996                        Xenocide (Ender's Saga  #3)            3   
997                                    Year of Wonders            4   
998                        You Suck (A Love Story  #2)            2   
999  Zen and the Art of Motorcycle Maintenance: An ...            4   

     avg_rating  
0      3.666667  
1      2.500000  
2      4.666667  
3      4.500000  
4      4.000000  
..          ...  
995    3.666667  
996    3.400000  
997    3.200000  
998    4.500000  
999    3.833333  

[1000 rows x 3 columns]

*Для каждой книги указаны количество обзоров и средняя оценка. Эти данные могут помочь определить, какие книги вызывают больше интереса у читателей (измеряется количеством обзоров) и какие книги имеют более высокие оценки от пользователей.*

**Задание 3: Определите издательство, которое выпустило наибольшее число книг толще 50 страниц:**

In [5]:
query = '''
SELECT
    p.publisher,
    COUNT(b.book_id) AS num_books
FROM publishers p
JOIN books b ON p.publisher_id = b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY num_books DESC
LIMIT 1;
'''
result_3 = pd.io.sql.read_sql(sql=text(query), con=con)
print("Задание 3:")
display(result_3)
print()

Задание 3:


publisher  num_books
0  Penguin Books         42

*Издательство с наибольшим числом книг, толще 50 страниц, является Penguin Books, выпустившим 42 таких книги. Это может быть полезной информацией для понимания предпочтений издательств и их активности.*

**Задание 4: Определите автора с самой высокой средней оценкой книг (с 50 и более оценками):**

In [10]:
query = '''
SELECT
    a.author,
    AVG(ra.rating) AS avg_rating
FROM authors a
JOIN (
    SELECT
        b.author_id,
        b.book_id
    FROM books b
    JOIN ratings ra ON b.book_id = ra.book_id
    GROUP BY b.author_id, b.book_id
    HAVING COUNT(ra.rating_id) >= 50
) books_with_fifty_ratings ON a.author_id = books_with_fifty_ratings.author_id
JOIN ratings ra ON books_with_fifty_ratings.book_id = ra.book_id
GROUP BY a.author_id, a.author
ORDER BY avg_rating DESC
LIMIT 1;
'''

# Выполняем SQL-запрос и выводим результат
result_4 = pd.io.sql.read_sql(sql=text(query), con=con)
print("Задание 4:")
display(result_4)
print()

Задание 4:


author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

*Автор с самой высокой средней оценкой книг, имеющих 50 и более оценок, - J.K. Rowling/Mary GrandPré. Её книги получили высокие оценки от пользователей, что может быть связано с их популярностью.*

**Задание 5: Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок:**

In [9]:
query = '''
SELECT AVG(subquery.reviews_per_user) AS avg_reviews_per_user
FROM (
    SELECT
        username,
        COUNT(DISTINCT review_id) AS reviews_per_user
    FROM reviews
    WHERE username IN (
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 48
    )
    GROUP BY username
) subquery;
'''
result_5 = pd.io.sql.read_sql(sql=text(query), con=con)
print("Задание 5:")
display(result_5)
print()

Задание 5:


avg_reviews_per_user
0                  24.0

*Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, составляет 24. Эта информация может дать представление о том, насколько активно пользователи оставляют обзоры, если они часто ставят оценки книгам.*